snapcraft/internal/sources/_git.py

```diff
@@ -15,14 +15,54 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 import os
+import re
 import subprocess
+import sys
 
 from . import errors
 from ._base import Base
 
 
 class Git(Base):
 
+    @classmethod
+    def generate_version(cls, *, source_dir=None):
+        """Return the latest git tag from PWD or defined source_dir.
+
+        The output depends on the use of annotated tags, an will return
+        something like: '2.28+git.10.abcdef' where '2.28 is the
+        tag, '+git' indicates there are commits ahead of the tag, in
+        this case it is '10' and the latest commit hash begins with
+        'abcdef'. If there are no tags or the revision cannot be
+        determined, this will return 0 as the tag and only the commit
+        hash of the latest commit.
+        """
+        if not source_dir:
+            source_dir = os.getcwd()
+
+        try:
+            output = subprocess.check_output([
+                'git', '-C', source_dir, 'describe',
+                '--dirty']).decode(sys.getfilesystemencoding()).strip()
+        except subprocess.CalledProcessError:
+            output = subprocess.check_output([
+                'git', '-C', source_dir, 'describe', '--tags', '--dirty',
+                '--always']).decode(sys.getfilesystemencoding()).strip()
+            return '0+git.{}'.format(output)
+
+        m = re.search(r'(?P<tag>.*)-(?P<revs_ahead>.*)-g(?P<commit>.*)',
```

Comment:
to avoid surprises I'd suggest anchoring this at both ends (i.e. start it with `^`, end it with `$`). Also if `revs_ahead` is expected to be a number, you want `\d+`, not `.*`. Also nothing here should be `.*`; you want at least one char, and they should all be non-whitespace probably (so `\S+`), and I suspect `commit` is expected to be hexadecimal so `[0-9a-fA-F]+`.